File: kotlin/src/main/kotlin/in/kanur/sorts/Quicksort.kt
package `in`.kanur.sorts

object Quicksort : Sort {
  data class Partition(var lower: IntArray, var higher: IntArray)

  override fun sort(list: IntArray): IntArray {
    return when (list.size) {
      0 -> list
      else -> {
        val pivot = list[list.size - 1]
        val partitions = list.sliceArray(IntRange(0, list.size - 2))
          .fold(Partition(intArrayOf(), intArrayOf())) { partitions, num ->
            if (num > pivot) {
              partitions.copy(higher = partitions.higher + intArrayOf(num))
            } else {
              partitions.copy(lower = partitions.lower + intArrayOf(num))
            }
          }

        sort(partitions.lower) + intArrayOf(pivot) + sort(partitions.higher)
      }
    }
  }
}

File: typescript/src/quicksort.ts
class Partition {
  public lower: number[] = [];
  public higher: number[] = [];
}

export function sort(numbers: number[]): number[] {
  const length = numbers.length;

  if (length === 0) {
    return numbers;
  }

  const pivot = numbers[numbers.length - 1];
  const partition = numbers.slice(0, numbers.length - 1)
    .reduce(
      (prtn: Partition, num: number) => {
        if (num > pivot) {
          return {...prtn, higher: [...prtn.higher, num]};
        }

        return {...prtn, lower: [...prtn.lower, num]};
      },
      new Partition()
    );

  return [
    ...sort(partition.lower),
    pivot,
    ...sort(partition.higher),
  ];
}
